### PR TITLE
feat: Enable adding files 

### PR DESCRIPTION
**Pull request by mattzh72**
## Description

Enable adding files (refactored from `documents`). Specifically, this PR:

- Defines data models for `File` objects
- Supports storing files in SQL databases
- Adds an API route to retrieve all the files associated with a source
- Add pagination to API route

Also, adds some items to the `.gitignore`. 

I ran alembic via the instructions:

```
poetry run alembic upgrade head
poetry run alembic revision --autogenerate -m "Add FileMetadata table and drop Documents table"
```

## Testing
Added a test that loads the `memgpt` paper, retrieves the documents via the new API route, and confirms that:

1. the correct number of files were made
2. the metadata contains the right file name
3. the sources are set correctly on the files

### CODE REVIEW

**Review by sarahwooders — CHANGES_REQUESTED** [review]
We may want to rename this concept to a file, and associate it with an entire PDF (or file more generally) that gets uploaded.

File: .gitignore [hunk 1]
```diff
@@ -2,6 +2,9 @@
 # Created by https://www.toptal.com/developers/gitignore/api/vim,linux,macos,pydev,python,eclipse,pycharm,windows,netbeans,pycharm+all,pycharm+iml,visualstudio,jupyternotebooks,visualstudiocode,xcode,xcodeinjection
 # Edit at https://www.toptal.com/developers/gitignore?templates=vim,linux,macos,pydev,python,eclipse,pycharm,windows,netbeans,pycharm+all,pycharm+iml,visualstudio,jupyternotebooks,visualstudiocode,xcode,xcodeinjection
 
+openapi_letta.json
```
[the remark this answers is not in the document]
Comment (mattzh72) [reply .gitignore:5]: We currently don't push this to the repo, so safe to add here.

File: letta/config.py [hunk 2]
```diff
@@ -78,6 +78,11 @@ class LettaConfig:
     metadata_storage_path: str = LETTA_DIR
     metadata_storage_uri: str = None
 
+    # database configs: document storage
```
[the remark this answers is not in the document]
Comment (mattzh72) [reply letta/config.py:81]: ack

File: letta/data_sources/connectors.py [hunk 3]
```diff
@@ -57,14 +57,14 @@ def load_data(
     for document_text, document_metadata in connector.generate_documents():
         # insert document into storage
         document = Document(
```
[the remark this answers is not in the document]
Comment (mattzh72) [reply letta/data_sources/connectors.py]: Ah got it, was mirroring Passage below.

File: letta/server/rest_api/routers/v1/sources.py [hunk 4]
```diff
@@ -186,18 +186,16 @@ def list_passages(
     return passages
 
 
-@router.get("/{source_id}/documents", response_model=List[Document], operation_id="list_source_documents")
-def list_documents(
+@router.get("/{source_id}/documents", response_model=List[Document], operation_id="list_documents_from_source")
```
[the remark this answers is not in the document]
Comment (mattzh72) [reply letta/server/rest_api/routers/v1/sources.py]: Will leave this to future PRs, doesn't seem needed by shub rn

[the remark this answers is not in the document]
Comment (mattzh72) [reply letta/server/rest_api/routers/v1/sources.py]: planning to add delete API in future PR as well

File: letta/agent_store/storage.py [hunk 5]
```diff
@@ -33,17 +33,17 @@ class TableType:
 
 # external data source tables
 PASSAGE_TABLE_NAME = "letta_passages"  # chunked/embedded passages (from source)
-DOCUMENT_TABLE_NAME = "letta_documents"  # original documents (from source)
+DOCUMENT_TABLE_NAME = "letta_documents"  # original files (from source)
```
Comment (sarahwooders) [inline letta/agent_store/storage.py:36]: delete?

Comment (mattzh72) [reply letta/agent_store/storage.py:36]: renamed

File: letta/schemas/file.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+from pydantic import BaseModel, Field
+
+from letta.schemas.letta_base import LettaBase
+from letta.utils import get_utc_time
+
+
+class FileBase(LettaBase):
+    """Base class for document schemas"""
+
+    __id_prefix__ = "doc"
```
Comment (sarahwooders) [inline letta/schemas/file.py:13]: rename to `doc`?

**Review by sarahwooders — APPROVED** [review]

File: examples/notebooks/data_connector.ipynb [hunk 7]
```diff
@@ -270,7 +270,7 @@
    "outputs": [],
    "source": [
     "from letta.data_sources.connectors import DataConnector \n",
-    "from letta.schemas.document import Document\n",
+    "from letta.schemas.document import File\n",
```
Comment (sarahwooders) [inline examples/notebooks/data_connector.ipynb:273]: `FileMetadata`?

Comment (mattzh72) [reply examples/notebooks/data_connector.ipynb:273]: ack

File: letta/schemas/file.py (hunk 6 above)
Comment (mattzh72) [reply letta/schemas/file.py:13]: ack good catch

File: letta/data_sources/connectors.py [hunk 8]
```diff
@@ -41,33 +43,33 @@ def load_data(
     connector: DataConnector,
     source: Source,
     passage_store: StorageConnector,
-    document_store: Optional[StorageConnector] = None,
+    document_store: StorageConnector,
```
[the remark this answers is not in the document]
Comment (mattzh72) [reply letta/data_sources/connectors.py]: ack

File: letta/client/client.py [hunk 9]
```diff
@@ -1088,6 +1098,30 @@ def list_attached_sources(self, agent_id: str) -> List[Source]:
             raise ValueError(f"Failed to list attached sources: {response.text}")
         return [Source(**source) for source in response.json()]
 
+    def list_files_from_source(self, source_id: str, limit: int = 10, cursor: Optional[str] = None) -> PaginatedListFilesResponse:
```
[the remark this answers is not in the document]
Comment (mattzh72) [reply letta/client/client.py:1101]: done

File: letta/data_sources/connectors.py [hunk 10]
```diff
@@ -1,35 +1,37 @@
...
         """
         Generate document text and metadata from a data source.
 
         Returns:
-            documents (Iterator[Tuple[str, Dict]]): Generate a tuple of string text and metadata dictionary for each document.
+            files (Iterator[Tuple[str, Dict]]): Generate a tuple of string text and metadata dictionary for each document.
         """
 
-    def generate_passages(self, documents: List[Document], chunk_size: int = 1024) -> Iterator[Tuple[str, Dict]]:  # -> Iterator[Passage]:
+    def generate_passages(self, file_text: str, file: File, chunk_size: int = 1024) -> Iterator[Tuple[str, Dict]]:  # -> Iterator[Passage]:
```
[the remark this answers is not in the document]
Comment (mattzh72) [reply letta/data_sources/connectors.py]: discussed offline, reshaped interface, see updated code

File: letta/data_sources/connectors.py [hunk 11]
```diff
@@ -143,59 +145,62 @@ def __init__(self, input_files: List[str] = None, input_directory: str = None, r
         if self.recursive == True:
             assert self.input_directory is not None, "Must provide input directory if recursive is True."
 
-    def generate_documents(self) -> Iterator[Tuple[str, Dict]]:  # -> Iterator[Document]:
+    def generate_files(self) -> Iterator[Tuple[str, Dict]]:
```
[the remark this answers is not in the document]
Comment (mattzh72) [reply letta/data_sources/connectors.py]: sounds good - discussed offline, basically user should take in last given file and pass that in as a cursor

File: letta/client/client.py [hunk 12]
```diff
@@ -2261,6 +2298,20 @@ def list_attached_sources(self, agent_id: str) -> List[Source]:
         """
         return self.server.list_attached_sources(agent_id=agent_id)
 
+    def list_files_from_source(self, source_id: str, limit: int = 10, cursor: Optional[str] = None) -> PaginatedListFilesResponse:
```
[the remark this answers is not in the document]
Comment (mattzh72) [reply letta/client/client.py:2301]: done

File: letta/client/client.py [hunk 13]
```diff
@@ -1088,6 +1098,30 @@ def list_attached_sources(self, agent_id: str) -> List[Source]:
...
+        params = {"limit": limit, "cursor": cursor}
+
+        # Make the request to the FastAPI endpoint
+        response = requests.get(f"{self.base_url}/{self.api_prefix}/sources/{source_id}/files", headers=self.headers, params=params)
+
+        if response.status_code != 200:
+            raise ValueError(f"Failed to list files with source id {source_id}: [{response.status_code}] {response.text}")
+
+        # Parse the JSON response
+        return PaginatedListFilesResponse(**response.json())
```
[the remark this answers is not in the document]
Comment (mattzh72) [reply letta/client/client.py:1123]: ack

**Comment by mattzh72** [comment]
cc @4shub for correctness - I noticed it didn't mention dropping the Documents table or adding a new FileMetadata table - is this expected?

**Comment by 4shub** [comment]
This is correct, alembic only does any change on database deletions or modifications. You did not remove any database you just removed a scheme

**Pull request merged**